x-pack/plugins/viz_editor/common/components/index_pattern_panel/index_pattern_panel.tsx

```diff
@@ -4,39 +4,88 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
-import { EuiButtonEmpty, EuiIcon, ICON_TYPES } from '@elastic/eui';
+import {
+  EuiIcon,
+  // @ts-ignore
+  EuiSuperSelect,
+  ICON_TYPES,
+} from '@elastic/eui';
 // @ts-ignore untyped dependency
 import { palettes } from '@elastic/eui/lib/services';
-import React, { useState } from 'react';
+import zipObject from 'lodash-es/zipObject';
+import React, { useEffect, useState } from 'react';
+import chrome from 'ui/chrome';
 import { IndexPatterns } from '../../lib';
-
-interface Props {
-  indexPatterns: IndexPatterns;
-}
+import { getIndexPatterns } from '../../lib/index_patterns';
 
 interface State {
-  indexPattern?: string;
+  selectedIndexName: string;
+  indexPatterns: IndexPatterns | null;
 }
 
-function initialState(indexPatterns: IndexPatterns): State {
+function initialState(): State {
+  const settingsClient = chrome.getUiSettingsClient();
   return {
-    indexPattern: Object.keys(indexPatterns)[0],
+    selectedIndexName: settingsClient.get('defaultIndex') || '',
+    indexPatterns: null,
   };
 }
 
-export function IndexPatternPanel({ indexPatterns }: Props) {
-  const [state] = useState(() => initialState(indexPatterns));
-  const indexPattern = state.indexPattern ? indexPatterns[state.indexPattern] : undefined;
+function getIndexPatternFromName(state: State, title: string) {
+  if (!state.indexPatterns) {
+    return null;
+  }
+
+  return Object.values(state.indexPatterns).find(indexPattern => {
+    return indexPattern.title === title;
+  });
+}
+
+export function IndexPatternPanel() {
+  const [state, setState] = useState(() => initialState());
 
-  if (!indexPattern) {
+  useEffect(() => {
```

Comment:
`useEffect` runs on every render. Not a big deal, as you have an internal condition that checks, but if we render multiple times while indexPatterns is unset, you'll be making multiple calls to getIndexPatterns. You can pass it a second argument, and then it'll only run if that argument has changed: https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects

I think we should be passing it `[state.indexPatterns]` or something like this.

Reply:
Thanks for the suggestion!